### PR TITLE
feat: implement fullscreen context + component

### PR DESCRIPTION
**Pull request by wadjih-bencheikh18**
closes https://github.com/zakodium-oss/analysis-ui-components/issues/217

### CODE REVIEW

**Comment by lpatiny** [comment]
In full screen I switch surprisingly to dark mode.

image

Also in the application the 'open stories' link should not be present because the full screen mode should be targeted to the following part:

image

**Comment by wadjih-bencheikh18** [comment]
> In full screen I switch surprisingly to dark mode.

I will try to find out whats the problem
Because i tried the functionality on other pages and it works perfectly

**Comment by wadjih-bencheikh18** [comment]
@lpatiny fullscreen bug fixed

File: src/app/App.tsx [hunk 1]
```diff
@@ -71,9 +78,19 @@ function DropZoneArea() {
         }}
       >
         <div
-          style={{ height: '100%', display: 'flex', flexDirection: 'column' }}
+          style={{
+            height: '100%',
+            display: 'flex',
+            flexDirection: 'column',
+            backgroundColor: 'white',
```
[the remark this answers is not in the document]
Comment (wadjih-bencheikh18) [reply src/app/App.tsx:85]: Because on full screen the background turns to black

File: src/components/context/FullscreenContext.tsx [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+      document.body.requestFullscreen().catch(() => {
+        alert('Fullscreen is not supported');
+      });
+    } else if (document.fullscreenElement !== null && document.exitFullscreen) {
+      document.exitFullscreen().catch(() => {
+        alert("Can't exit fullscreen");
+      });
+    }
+  }, [isFullScreen]);
+  return <>{children}</>;
```
Comment (stropitek) [inline src/components/context/FullscreenContext.tsx:61]: Create a div here, and use its ref to request full screen on it (not on `document.body`)

The div can have 100% width and height, and a white background

**Comment by stropitek** [comment]
Please create an example with 2 nested parts of the UI which can be toggled to fullscreen.

**Review by stropitek — COMMENTED** [review]
Please create an example with 2 nested parts of the UI which can be toggled to fullscreen.

File: src/app/App.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (targos) [reply src/app/App.tsx:85]: Ok I found some references. Please remove the backgroundColor from here and add it where @stropitek  commented.

File: src/components/context/FullscreenContext.tsx (hunk 2 above)
Comment (wadjih-bencheikh18) [reply src/components/context/FullscreenContext.tsx]: fixed thank you

File: src/app/App.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (wadjih-bencheikh18) [reply src/app/App.tsx]: done, thank you

**Comment by wadjih-bencheikh18** [comment]
> I meant 1 example which nests 2 fullscreen elements, not 2 examples which each have one fullscreen element.

done, sorry for the misunderstanding

**Comment by stropitek** [comment]
"Nest" means 1 contains the other 😅 

```jsx
<FullScreenProvider>
  <Content i="1" />
  <FullScreenProvider>  
    <Content i="2" />
  </FullScreenProvider>
</FullScreenProvider>
```

**Comment by targos** [comment]
^ @stropitek do you expect to be able to nest the fullscreen action? Meaning if you put the parent in full screen, you can then put the child too, ending up with two elements in full screen.

**Comment by wadjih-bencheikh18** [comment]
> "Nest" means 1 contains the other

Implemented.

**Comment by stropitek** [comment]
> do you expect to be able to nest the fullscreen action? Meaning if you put the parent in full screen, you can then put the child too, ending up with two elements in full screen.

I see nesting as a potentially valid use case (for example full app + spectra area) and I wanted to see how it behaves with the current implementation. Apparently it does not behave correctly, but I'm not expecting that it works in this PR.

I guess we need to validate that nesting / stacking is a valid use case. @lpatiny any thoughts?

**Comment by lpatiny** [comment]
I think it make sense that many parts of the same application can go full screen and they could be nested but I think if we are in full screen we should not be able to put in full screen a part of it. This means that all the 'full screen' button should disappear once we are in full screen.

I'm not aware of any application in which we can have a stack of full screen and this could be quite confusing for the end user.

**Comment by targos** [comment]
This behavior is weird if you "Toggle fullscreen 1" and the try to "Toggle fullscreen 2".

>  I think if we are in full screen we should not be able to put in full screen a part of it.

Why not? I think it should either replace the existing full screen element or create a stack, but not remove buttons (that would be difficult to implement)

**Comment by wadjih-bencheikh18** [comment]
> This behavior is weird if you "Toggle fullscreen 1" and the try to "Toggle fullscreen 2".

I fixed the behavior and it works perfectly now

**Comment by targos** [comment]
I think there's still an issue when the user exits fullscreen using the Escape key. In that case, you need to click twice on the button to enter fullscreen again.

**Pull request merged**